### PR TITLE
Issue 307 - Webpack build plugin

### PR DESCRIPTION
**Pull request by Hardekerlis**

### CODE REVIEW

**Comment by Ole1288** [comment]
Need better error messages with colors and stuff.

These aint okay:
Failed to upload sourcemaps
{
  "errors": [
    {
      "code": 4001,
      "message": "Not authorized"
    }
  ]
}
----------------------------------------------------------------------------------------------
SyntaxError: Unexpected token < in JSON at position 0
    at JSON.parse (<anonymous>)
    at Response.json (file:///home/skycake/webpack-testing-folder/byteboost-webpack-sourcemap-plugin/node_modules/node-fetch/src/body.js:149:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async UploadSourceMapsHandler.uploadSourcemaps (file:///home/skycake/webpack-testing-folder/byteboost-webpack-sourcemap-plugin/node_modules/@byteboost/cli/dist/managers/sourcemaps/upload.js:144:40)
    at async file:///home/skycake/webpack-testing-folder/byteboost-webpack-sourcemap-plugin/dist/index.js:81:13

File: src/index.ts [hunk 1]
```diff
@@ -1,55 +1,141 @@
...
+    if (!options.version && !options.versionFromPackageJSON) {
+      throw new Error('version is required');
+    }
+
+    if (options.version && options.versionFromPackageJSON) {
+      throw new Error(
+        'version and versionFromPackageJSON cannot be used together',
+      );
+    }
+
```
Comment (Ole1288) [inline src/index.ts]: Should display that these error are from the byteboost plugin

e..g prefix with byteboost or something

File: src/index.ts [hunk 2]
```diff
@@ -1,55 +1,141 @@
...
+    try {
+      const packageJSONContent = readFileSync(
+        `${this.compiler?.context || ''}/package.json`,
+        'utf-8',
+      );
+
+      const packageJSON = this.parseJSON(packageJSONContent);
+
+      if (!packageJSON) {
+        throw new Error('Couldnt parse package.json');
```
Comment (Ole1288) [inline src/index.ts]: same

File: src/index.ts [hunk 3]
```diff
@@ -1,55 +1,141 @@
...
+  async apply(compiler: Webpack.Compiler) {
+    this.compiler = compiler;
+
+    const handler = new UploadSourceMapsHandler(compiler.context);
+
+    const isValidJsDir = handler.isPathValidJsDirectory();
+
+    if (!isValidJsDir) {
+      this.warn(
+        `The path ${path} is not a valid JS project directory. Couldn't find package.json`,
```
Comment (Ole1288) [inline src/index.ts:101]: same

File: src/index.ts [hunk 4]
```diff
@@ -1,55 +1,141 @@
...
+      if (!handler.mapFilePaths[0]) {
+        this.warn(`No sourcemaps found in ${handler.fullpath}`);
+        return;
+      }
+
+      await handler.tagFilesWithDebugInfo();
+
+      await handler.uploadSourcemaps();
+
+      this.log(`Uploaded sourcemaps.`);
```
Comment (Ole1288) [inline src/index.ts:134]: same

File: src/index.ts [hunk 5]
```diff
@@ -1,55 +1,141 @@
...
+      BYTEBOOST_ORGANIZATION: this.options.organization,
+      BYTEBOOST_DOMAIN: this.options.domain,
+      BYTEBOOST_TOKEN: this.options.token,
+    };
+
+    compiler.hooks.afterDone.tap('ByteboostSourcemap', async () => {
+      handler.compileSourceMapPathsList();
+
+      if (!handler.mapFilePaths[0]) {
+        this.warn(`No sourcemaps found in ${handler.fullpath}`);
```
Comment (Ole1288) [inline src/index.ts:120]: same

File: src/index.ts (hunk 4 above)
Comment (Ole1288) [inline src/index.ts:134]: same

**Comment by Hardekerlis** [comment]
> Need better error messages with colors and stuff.
> 
> ## These aint okay:
> Failed to upload sourcemaps
> {
> "errors": [
> {
> "code": 4001,
> "message": "Not authorized"
> }
> ]
> }
> SyntaxError: Unexpected token < in JSON at position 0 at JSON.parse () at Response.json (file:///home/skycake/webpack-testing-folder/byteboost-webpack-sourcemap-plugin/node_modules/node-fetch/src/body.js:149:15) at process.processTicksAndRejections (node:internal/process/task_queues:95:5) at async UploadSourceMapsHandler.uploadSourcemaps (file:///home/skycake/webpack-testing-folder/byteboost-webpack-sourcemap-plugin/node_modules/@byteboost/cli/dist/managers/sourcemaps/upload.js:144:40) at async file:///home/skycake/webpack-testing-folder/byteboost-webpack-sourcemap-plugin/dist/index.js:81:13

I see your point but I dont think it is necessary to complicate error handling atm. I agree it should be prettier in the future but I do think what went wrong is pretty clear

**Pull request merged**